overlord/snapstate/snapstate.go

```diff
@@ -1876,17 +1880,35 @@ type update struct {
 	Components []ComponentSetup
 }
 
-// revisionSatisfied returns true if the state of the snap on the system matches the
-// state specified in the update. This method is primarily concerned with the
-// revision of the snap.
-//
-// TODO:COMPS: check if we need to change the state of components
-func (u *update) revisionSatisfied() bool {
+// revisionSatisfied returns true if the state of the snap on the system matches
+// the state specified in the update. This method checks if the currently
+// installed snap and components match what is specified in the update.
+func (u *update) revisionSatisfied() (bool, error) {
 	if u.Setup.AlwaysUpdate || !u.SnapState.IsInstalled() {
-		return false
+		return false, nil
+	}
+
+	if u.SnapState.Current != u.Setup.Revision() {
+		return false, nil
+	}
+
+	comps, err := u.SnapState.CurrentComponentInfos()
+	if err != nil {
+		return false, err
+	}
+
+	currentCompRevs := make(map[string]snap.Revision, len(comps))
+	for _, comp := range comps {
+		currentCompRevs[comp.Component.ComponentName] = comp.Revision
 	}
 
-	return u.SnapState.Current == u.Setup.Revision()
+	for _, comp := range u.Components {
+		if currentCompRevs[comp.CompSideInfo.Component.ComponentName] != comp.Revision() {
+			return false, nil
+		}
+	}
```

Comment:
You're right that it shouldn't be possible, since a snap's snap.yaml should define the available components, and that can't change without a revision change. Do you think that a consistency check is enough? I'd rather not write the code for a case that shouldn't be possible.

Reply:
Note, we already do error out if the resource we're trying to install isn't available in the store.

Reply:
Ok, that should be enough, thanks